interp/constrextern.ml

```diff
@@ -63,6 +63,27 @@ let print_universes = Detyping.print_universes
 (* This suppresses printing of primitive tokens (e.g. numeral) and notations *)
 let print_no_symbol = ref false
 
+(* This tells to skip types if a variable has this type by default *)
+let print_use_implicit_types =
+  Goptions.declare_bool_option_and_ref
+    ~depr:false
+    ~name:"use of implicit types in printing"
+    ~key:["Printing";"Use";"Implicit";"Types"]
+    ~value:true
+
+(**********************************************************************)
+
+let skip_if_reserved_type na t =
+  if !Flags.raw_print || not (print_use_implicit_types ()) then t else
+  match na with
+  | Anonymous -> t
+  | Name id ->
+    try
+      let pat = Reserve.find_reserved_type id in
+      let _ = match_notation_constr false t ([],pat) in
```

Comment:
BTW I don't know much about this function, will it match holes like
~~~coq
Implicit Types b : bool.

(* let x := _ forces an evar instead of getting bool *)
Eval compute in let x := _ in forall b1 b2 : x, b1 = b2.
(*      = forall b1 b2 : ?T, b1 = b2 *)
~~~
?

Reply:
It doesn't match. A `_` in the notation would match any term but a `_` of the term is not matched by a concrete term of the notation.

This actually raises the question of interaction between dropping implicit types and the notation mechanism. The answer is that a notation is first searched. For instance, if `n` is bound to `nat` and I have a notation say, for `forall p q:nat, p + q = 0`, and the notation is applied to `forall n1 n2 : nat, n1 + n2 = 0`, the notation is tried first (and the implicit type removal does not occur).